assets/shiny-server.js

```diff
@@ -325,70 +325,149 @@
-          clearInterval(countdown);
-          self._doClose();
+
+        self._openConnection_p()
+        .then(function(){
           $('body').removeClass('ss-reconnecting');
+          $('#ss-connect-dialog').remove();
+          $('#ss-gray-out').remove();
+          def.resolve();
+        }, function(err){
+          // This was a failed attempt to reconnect
+          log("Unable to reconnect: " + JSON.stringify(err));
+          def.reject();
+        });
 
-          $('#ss-connect-dialog').html('<button id="ss-reload-button" type="button" class="ss-dialog-button">Reload</button> Disconnected from the server.');
-          $('#ss-reload-button').click(function(){
-            location.reload();
-          });
+        return def;
+      }
+
+      // @param time The last time a connection attempt was started
+      // @param count 0-indexed count of how many reconnect attempts have occured.
+      // @param expires the time when the session is scheduled to expire on 
+      // the server.
+      // @param timer The setTimeout object from a previous caller that was 
+      //   being used to update a dialog. This should be cleared before
+      //   starting a new timer.
+      function scheduleReconnect_p(time, count, expires, timer) {
+        var def = $.Deferred();
+
+        if (Date.now() > expires){
+          // Shouldn't happen, but if the current time is after the known
+          // expiration for this session, give up.
+          debug('Overshot session expiration.');
+          return def.reject(new Error("Overshot session expiration"));
+        }
+
+        // Compute delay exponentially.
+        var interval;
+        if (count < 10){
+          interval = 1000 * Math.pow(2, count);
+          interval = Math.min(15 * 1000, interval); // Max of 15s delay.
+        } else { 
+          // The interval may end up being configurable or changed, so we don't cut off
+          // exactly at 2^4, but don't bother computing really large powers.
+          interval = 15 * 1000;
+        }
+        var delay = time - Date.now() + interval;
+
+        // If the next attempt would be after the session is due to expire, 
+        // schedule one last attempt to connect a couple seconds before the
+        // expiration.
+        if (Date.now() + delay > (expires - 2000)) {
+          delay = expires - Date.now() - 2000;
+        }
+        if (delay < 0){
+          // i.e. we're within 2 seconds of session expiration. Make
+          // one last connection attmpt but don't schedule any more.
+          debug('Final reconnection attempt');
+          return reconnect_p();
+        }
+
+        if (timer){
+          clearTimeout(timer);
         }
+        var targetTime = Date.now() + delay;
+        function updateMsg() {
+          setReconnectDialog('Disconnected from server. Going to reconnect in ' + Math.ceil((targetTime-Date.now())/1000) + 's');
+        }
+        var msgTimer = setInterval(updateMsg, 500);
```

Comment:
A followup to my other message about this `msgTimer`. It really should be canceled immediately before you call `reconnect_p`. That way, you don't need to worry about passing it through the recursion. In other words: Tick until you are about to reconnect and then cancel the ticker. Don't wait for reconnect success/failure.

This is more correct because your current timer may trigger while the reconnect is happening and incorrect update the reconnect dialog.